pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/RealtimeSegmentAssignment.java

```diff
@@ -260,13 +277,26 @@ public Map<String, Map<String, String>> rebalanceTable(Map<String, Map<String, S
           "No COMPLETED instance partitions found, reassigning COMPLETED segments the same way as CONSUMING segments "
               + "with CONSUMING instance partitions for table: {}", _realtimeTableName);
 
+      Set<String> uploadedSegmentsWithNoPartitionMetadata = new TreeSet<>();
       newAssignment = new TreeMap<>();
       for (String segmentName : completedSegmentAssignment.keySet()) {
+        if (!SegmentUtils
+            .getRealtimeSegmentPartitionId(segmentName, _realtimeTableName, _helixManager, _partitionColumn)
+            .isPresent()) {
+          uploadedSegmentsWithNoPartitionMetadata.add(segmentName);
+          continue;
+        }
         List<String> instancesAssigned = assignConsumingSegment(segmentName, consumingInstancePartitions);
         Map<String, String> instanceStateMap =
             SegmentAssignmentUtils.getInstanceStateMap(instancesAssigned, SegmentStateModel.ONLINE);
         newAssignment.put(segmentName, instanceStateMap);
       }
+      for (String segmentName : uploadedSegmentsWithNoPartitionMetadata) {
```

Comment:
We should not assign one by one here. We may create a new map `unpartitionedCompletedSegmentAssignment` with only the unpartitioned segments, then use `reassignSegments()` to get minimum data movement